File: 10-code-control/15-prerequisite-flags/go/main.go
// Console grid navigator demonstrating LaunchDarkly flag prerequisites.
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
	"github.com/launchdarkly/go-sdk-common/v3/ldreason"
	ld "github.com/launchdarkly/go-server-sdk/v7"
	"golang.org/x/term"
)

// LaunchDarkly: evaluate the dependent flag even when its prerequisite fails.
// https://launchdarkly.com/docs/home/flags/prereqs
const (
	flagHighlight = "enable-grid-selection-highlight-prereq"
	flagCount     = "show-navigation-move-count-prereq"
	appBanner     = "15-prerequisite-flags[go]"
	bgANSI        = "\033[48;5;236m"
	resetANSI     = "\033[0m"
)

var (
	rows     = [3]string{"t", "m", "b"}
	cols     = [3]string{"l", "m", "r"}
	ldClient *ld.LDClient
)

type flagValues struct {
	username       string
	highlightColor string
	showMoveCount  bool
	parentValue    string
	parentReason   string
	childValue     bool
	childReason    string
}

type position struct {
	row, col int
}

type moveResult struct {
	row, col int
	moved    bool
}

type keyEvent struct {
	dr, dc      int
	action      sessionAction
	endsSession bool
	hasMove     bool
}

type sessionAction int

const (
	actionQuit sessionAction = iota
	actionLogout
)

func initLaunchDarkly() {
	sdkKey := os.Getenv("LD_SDK_KEY")
	if sdkKey == "" {
		fmt.Fprintln(os.Stderr, "Warning: LD_SDK_KEY not set — flags use safe defaults.")
		return
	}
	client, err := ld.MakeClient(sdkKey, 5*time.Second)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Warning: LaunchDarkly SDK did not initialize.")
		return
	}
	ldClient = client
}

func formatReason(reason ldreason.EvaluationReason) string {
	kind := string(reason.GetKind())
	if reason.GetKind() == ldreason.EvalReasonPrerequisiteFailed {
		if key := reason.GetPrerequisiteKey(); key != "" {
			return kind + " (" + key + ")"
		}
	}
	return kind
}

func highlightColor(value string) string {
	switch value {
	case "green", "yellow", "red", "blue", "purple", "pink":
		return value
	default:
		return "none"
	}
}

func evaluateFlags(username string) flagValues {
	userKey := strings.ToLower(strings.TrimSpace(username))
	if ldClient == nil {
		return flagValues{
			username:       userKey,
			highlightColor: "none",
			parentValue:    "none",
			parentReason:   "OFFLINE",
			childReason:    "OFFLINE",
		}
	}
	ctx := ldcontext.NewBuilder(userKey).Kind("user").Build()
	parentValue, parentDetail, _ := ldClient.StringVariationDetail(flagHighlight, ctx, "none")
	childValue, childDetail, _ := ldClient.BoolVariationDetail(flagCount, ctx, false)
	return flagValues{
		username:       userKey,
		highlightColor: highlightColor(parentValue),
		showMoveCount:  childValue,
		parentValue:    parentValue,
		parentReason:   formatReason(parentDetail.Reason),
		childValue:     childValue,
		childReason:    formatReason(childDetail.Reason),
	}
}

func ansiColor(color string) string {
	switch color {
	case "pink":
		return "\033[95m"
	case "yellow":
		return "\033[93m"
	case "red":
		return "\033[91m"
	case "blue":
		return "\033[94m"
	case "green":
		return "\033[92m"
	case "purple":
		return "\033[35m"
	default:
		return ""
	}
}

func colorize(text, color string) string {
	code := ansiColor(color)
	if code == "" {
		return text
	}
	return code + text + resetANSI + bgANSI
}

func formatPos(row, col int) string {
	return rows[row] + "/" + cols[col]
}

func tryMove(row, col, dr, dc int) moveResult {
	nr := clamp(row+dr, 0, 2)
	nc := clamp(col+dc, 0, 2)
	return moveResult{nr, nc, nr != row || nc != col}
}

func clamp(value, low, high int) int {
	if value < low {
		return low
	}
	if value > high {
		return high
	}
	return value
}

type readState int

const (
	keyReady readState = iota
	keyTimeout
	keyClosed
)

type keyStream struct {
	keys chan byte
}

func newKeyStream() *keyStream {
	stream := &keyStream{keys: make(chan byte, 64)}
	go func() {
		defer close(stream.keys)
		reader := bufio.NewReader(os.Stdin)
		for {
			key, err := reader.ReadByte()
			if err != nil {
				return
			}
			stream.keys <- key
		}
	}()
	return stream
}

func (stream *keyStream) next(timeout time.Duration) (byte, readState) {
	timer := time.NewTimer(timeout)
	defer timer.Stop()
	select {
	case key, open := <-stream.keys:
		if !open {
			return 0, keyClosed
		}
		return key, keyReady
	case <-timer.C:
		return 0, keyTimeout
	}
}

func (stream *keyStream) readLine() (string, bool) {
	var line strings.Builder
	for key := range stream.keys {
		if key == '\n' || key == '\r' {
			return strings.TrimSpace(line.String()), true
		}
		line.WriteByte(key)
	}
	return "", false
}

func readUsername(stream *keyStream) (string, error) {
	fmt.Println(appBanner)
	fmt.Println("Login")
	fmt.Println()
	for {
		fmt.Print("Username: ")
		line, ok := stream.readLine()
		if !ok {
			return "", fmt.Errorf("stdin closed")
		}
		if line != "" {
			return line, nil
		}
		fmt.Println("Username is required.")
	}
}

func drawCell(selected bool, color string, line int) string {
	var text string
	if selected {
		text = [3]string{"┏━━━┓", "┃ X ┃", "┗━━━┛"}[line]
		if color != "none" {
			return colorize(text, color)
		}
		return text
	}
	return [3]string{"┌───┐", "│   │", "└───┘"}[line]
}

func writeLine(out *strings.Builder, line string) {
	out.WriteString(line)
	out.WriteString("\033[K\r\n")
}

func render(username string, row, col int, previous *position, moveCount int, flags flagValues) {
	var out strings.Builder
	out.WriteString(bgANSI)
	out.WriteString("\033[H\033[2J")
	previousText := "—"
	if previous != nil {
		previousText = formatPos(previous.row, previous.col)
	}
	writeLine(&out, appBanner)
	writeLine(&out, "Name: "+colorize(username, flags.highlightColor))
	writeLine(&out, "Current position: "+formatPos(row, col))
	writeLine(&out, "Previous position: "+previousText)
	if flags.showMoveCount {
		writeLine(&out, fmt.Sprintf("Count: %d", moveCount))
	}
	writeLine(&out, "")
	writeLine(&out, fmt.Sprintf("Parent: %v  %s", flags.parentValue, flags.parentReason))
	writeLine(&out, fmt.Sprintf("Child:  %v  %s", flags.childValue, flags.childReason))
	writeLine(&out, "")
	writeLine(&out, "Use arrow keys or WASD to move (L to logout, Q to quit).")
	writeLine(&out, "")
	for r := 0; r < 3; r++ {
		for line := 0; line < 3; line++ {
			cells := make([]string, 3)
			for c := 0; c < 3; c++ {
				color := "none"
				if r == row && c == col {
					color = flags.highlightColor
				}
				cells[c] = drawCell(r == row && c == col, color, line)
			}
			writeLine(&out, strings.Join(cells, " "))
		}
	}
	fmt.Print(out.String())
}

func readKeyEvent(stream *keyStream, timeout time.Duration) keyEvent {
	key, state := stream.next(timeout)
	switch state {
	case keyTimeout:
		return keyEvent{}
	case keyClosed:
		return keyEvent{action: actionQuit, endsSession: true}
	}
	switch key {
	case 3, 'q', 'Q':
		return keyEvent{action: actionQuit, endsSession: true}
	case 'l', 'L':
		return keyEvent{action: actionLogout, endsSession: true}
	case 'w', 'W':
		return keyEvent{dr: -1, hasMove: true}
	case 's', 'S':
		return keyEvent{dr: 1, hasMove: true}
	case 'a', 'A':
		return keyEvent{dc: -1, hasMove: true}
	case 'd', 'D':
		return keyEvent{dc: 1, hasMove: true}
	case 27:
		second, state := stream.next(50 * time.Millisecond)
		if state != keyReady || second != '[' {
			return keyEvent{}
		}
		third, state := stream.next(50 * time.Millisecond)
		if state != keyReady {
			return keyEvent{}
		}
		switch third {
		case 'A':
			return keyEvent{dr: -1, hasMove: true}
		case 'B':
			return keyEvent{dr: 1, hasMove: true}
		case 'C':
			return keyEvent{dc: 1, hasMove: true}
		case 'D':
			return keyEvent{dc: -1, hasMove: true}
		}
	}
	return keyEvent{}
}

func runGrid(username string, stream *keyStream) sessionAction {
	row, col := 1, 1
	var previous *position
	moveCount := 0
	for {
		flags := evaluateFlags(username)
		render(flags.username, row, col, previous, moveCount, flags)
		event := readKeyEvent(stream, 500*time.Millisecond)
		if event.endsSession {
			return event.action
		}
		if !event.hasMove {
			continue
		}
		result := tryMove(row, col, event.dr, event.dc)
		if result.moved {
			old := position{row, col}
			previous = &old
			row, col = result.row, result.col
			moveCount++
		}
	}
}

func main() {
	initLaunchDarkly()
	defer func() {
		if ldClient != nil {
			ldClient.Close()
		}
		fmt.Print(resetANSI)
	}()

	stream := newKeyStream()
	fd := int(os.Stdin.Fd())
	for {
		username, err := readUsername(stream)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		oldState, err := term.MakeRaw(fd)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		action := runGrid(username, stream)
		_ = term.Restore(fd, oldState)
		fmt.Print(resetANSI)
		if action == actionQuit {
			return
		}
	}
}

File: 10-code-control/15-prerequisite-flags/rust/src/main.rs
//! Console grid navigator demonstrating LaunchDarkly flag prerequisites.

use crossterm::{
    cursor::{Hide, MoveTo, Show},
    event::{self, Event, KeyCode, KeyEvent, KeyModifiers},
    execute, queue,
    style::Print,
    terminal::{self, ClearType},
};
use launchdarkly_server_sdk::{Client, ConfigBuilder, ContextBuilder, Reason};
use std::io::{self, Write};
use std::sync::Arc;
use std::time::Duration;

// LaunchDarkly: evaluate the dependent flag even when its prerequisite fails.
// https://launchdarkly.com/docs/home/flags/prereqs
const FLAG_HIGHLIGHT: &str = "enable-grid-selection-highlight-prereq";
const FLAG_COUNT: &str = "show-navigation-move-count-prereq";
const APP_BANNER: &str = "15-prerequisite-flags[rust]";
const BG: &str = "\x1b[48;5;236m";
const RESET: &str = "\x1b[0m";
const ROWS: [&str; 3] = ["t", "m", "b"];
const COLS: [&str; 3] = ["l", "m", "r"];

struct Position {
    row: i32,
    col: i32,
}

struct MoveResult {
    row: i32,
    col: i32,
    moved: bool,
}

#[derive(PartialEq, Eq)]
enum SessionAction {
    Quit,
    Logout,
}

struct FlagValues {
    username: String,
    highlight_color: String,
    show_move_count: bool,
    parent_value: String,
    parent_reason: String,
    child_value: bool,
    child_reason: String,
}

struct App {
    client: Option<Arc<Client>>,
}

fn format_reason(reason: &Reason) -> String {
    match reason {
        Reason::Off => "OFF".to_string(),
        Reason::Fallthrough { .. } => "FALLTHROUGH".to_string(),
        Reason::TargetMatch => "TARGET_MATCH".to_string(),
        Reason::RuleMatch { .. } => "RULE_MATCH".to_string(),
        Reason::PrerequisiteFailed { prerequisite_key } => {
            format!("PREREQUISITE_FAILED ({prerequisite_key})")
        }
        Reason::Error { .. } => "ERROR".to_string(),
    }
}

fn highlight_color(value: &str) -> String {
    match value {
        "green" | "yellow" | "red" | "blue" | "purple" | "pink" => value.to_string(),
        _ => "none".to_string(),
    }
}

impl App {
    fn new() -> Self {
        let sdk_key = std::env::var("LD_SDK_KEY").ok();
        let client = sdk_key.and_then(|key| {
            let config = ConfigBuilder::new(&key).build().ok()?;
            let client = Client::build(config).ok()?;
            client.start_with_runtime().ok()?;
            let deadline = std::time::Instant::now() + Duration::from_secs(5);
            while !client.initialized() && std::time::Instant::now() < deadline {
                std::thread::sleep(Duration::from_millis(50));
            }
            if client.initialized() {
                Some(Arc::new(client))
            } else {
                eprintln!("Warning: LaunchDarkly SDK did not initialize.");
                None
            }
        });
        if std::env::var("LD_SDK_KEY").is_err() {
            eprintln!("Warning: LD_SDK_KEY not set — flags use safe defaults.");
        }
        Self { client }
    }

    fn evaluate_flags(&self, username: &str) -> FlagValues {
        let user_key = username.trim().to_lowercase();
        let Some(client) = &self.client else {
            return FlagValues {
                username: user_key,
                highlight_color: "none".to_string(),
                show_move_count: false,
                parent_value: "none".to_string(),
                parent_reason: "OFFLINE".to_string(),
                child_value: false,
                child_reason: "OFFLINE".to_string(),
            };
        };
        let context = ContextBuilder::new(&user_key)
            .kind("user")
            .build()
            .unwrap_or_else(|_| ContextBuilder::new("anonymous").build().unwrap());
        let parent = client.str_variation_detail(&context, FLAG_HIGHLIGHT, "none".to_string());
        let child = client.bool_variation_detail(&context, FLAG_COUNT, false);
        let parent_value = parent.value.unwrap_or_else(|| "none".to_string());
        let child_value = child.value.unwrap_or(false);
        FlagValues {
            username: user_key,
            highlight_color: highlight_color(&parent_value),
            show_move_count: child_value,
            parent_value,
            parent_reason: format_reason(&parent.reason),
            child_value,
            child_reason: format_reason(&child.reason),
        }
    }
}

impl Drop for App {
    fn drop(&mut self) {
        if let Some(client) = self.client.take() {
            if let Ok(client) = Arc::try_unwrap(client) {
                client.close();
            }
        }
    }
}

fn ansi_color(color: &str) -> &'static str {
    match color {
        "pink" => "\x1b[95m",
        "yellow" => "\x1b[93m",
        "red" => "\x1b[91m",
        "blue" => "\x1b[94m",
        "green" => "\x1b[92m",
        "purple" => "\x1b[35m",
        _ => "",
    }
}

fn colorize(text: &str, color: &str) -> String {
    let ansi = ansi_color(color);
    if ansi.is_empty() {
        text.to_string()
    } else {
        format!("{ansi}{text}{RESET}{BG}")
    }
}

fn format_pos(row: i32, col: i32) -> String {
    format!("{}/{}", ROWS[row as usize], COLS[col as usize])
}

fn try_move(row: i32, col: i32, dr: i32, dc: i32) -> MoveResult {
    let nr = (row + dr).clamp(0, 2);
    let nc = (col + dc).clamp(0, 2);
    if nr == row && nc == col {
        MoveResult { row, col, moved: false }
    } else {
        MoveResult {
            row: nr,
            col: nc,
            moved: true,
        }
    }
}

fn read_username() -> io::Result<String> {
    println!("{APP_BANNER}");
    println!("Login\n");
    loop {
        print!("Username: ");
        io::stdout().flush()?;
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        let name = line.trim().to_string();
        if !name.is_empty() {
            return Ok(name);
        }
        println!("Username is required.");
    }
}

fn cell_line(selected: bool, color: &str, line: usize) -> String {
    let plain = match line {
        0 => "┏━━━┓",
        1 => "┃ X ┃",
        _ => "┗━━━┛",
    };
    if selected {
        if !color.is_empty() && color != "none" {
            colorize(plain, color)
        } else {
            plain.to_string()
        }
    } else {
        match line {
            0 => "┌───┐".to_string(),
            1 => "│   │".to_string(),
            _ => "└───┘".to_string(),
        }
    }
}

fn print_line(out: &mut impl Write, y: u16, text: &str) -> io::Result<()> {
    queue!(out, MoveTo(0, y), Print(text))?;
    Ok(())
}

fn render(
    out: &mut impl Write,
    username: &str,
    row: i32,
    col: i32,
    previous: Option<&Position>,
    move_count: i32,
    flags: &FlagValues,
) -> io::Result<()> {
    execute!(out, MoveTo(0, 0), terminal::Clear(ClearType::All))?;
    queue!(out, Print(BG))?;
    let prev_text = previous
        .map(|p| format_pos(p.row, p.col))
        .unwrap_or_else(|| "—".to_string());
    let mut y = 0u16;
    print_line(out, y, APP_BANNER)?;
    y += 1;
    print_line(out, y, &format!("Name: {}", colorize(username, &flags.highlight_color)))?;
    y += 1;
    print_line(out, y, &format!("Current position: {}", format_pos(row, col)))?;
    y += 1;
    print_line(out, y, &format!("Previous position: {prev_text}"))?;
    if flags.show_move_count {
        y += 1;
        print_line(out, y, &format!("Count: {move_count}"))?;
    }
    y += 2;
    print_line(
        out,
        y,
        &format!("Parent: {}  {}", flags.parent_value, flags.parent_reason),
    )?;
    y += 1;
    print_line(
        out,
        y,
        &format!("Child:  {}  {}", flags.child_value, flags.child_reason),
    )?;
    y += 2;
    print_line(out, y, "Use arrow keys or WASD to move (L to logout, Q to quit).")?;
    y += 2;
    for r in 0..3 {
        let top = (0..3)
            .map(|c| {
                let selected = r == row && c == col;
                let color = if selected {
                    flags.highlight_color.as_str()
                } else {
                    "none"
                };
                cell_line(selected, color, 0)
            })
            .collect::<Vec<_>>()
            .join(" ");
        let mid = (0..3)
            .map(|c| {
                let selected = r == row && c == col;
                let color = if selected {
                    flags.highlight_color.as_str()
                } else {
                    "none"
                };
                cell_line(selected, color, 1)
            })
            .collect::<Vec<_>>()
            .join(" ");
        let bot = (0..3)
            .map(|c| {
                let selected = r == row && c == col;
                let color = if selected {
                    flags.highlight_color.as_str()
                } else {
                    "none"
                };
                cell_line(selected, color, 2)
            })
            .collect::<Vec<_>>()
            .join(" ");
        print_line(out, y, &top)?;
        y += 1;
        print_line(out, y, &mid)?;
        y += 1;
        print_line(out, y, &bot)?;
        y += 1;
    }
    out.flush()?;
    Ok(())
}

fn run_grid(out: &mut impl Write, app: &App, username: &str) -> io::Result<SessionAction> {
    let mut row = 1;
    let mut col = 1;
    let mut previous: Option<Position> = None;
    let mut move_count = 0;
    loop {
        let flags = app.evaluate_flags(username);
        render(
            out,
            &flags.username,
            row,
            col,
            previous.as_ref(),
            move_count,
            &flags,
        )?;
        if !event::poll(Duration::from_millis(500))? {
            continue;
        }
        if let Event::Key(KeyEvent { code, modifiers, .. }) = event::read()? {
            if modifiers.contains(KeyModifiers::CONTROL) {
                return Ok(SessionAction::Quit);
            }
            let movement = match code {
                KeyCode::Char('q') | KeyCode::Char('Q') => return Ok(SessionAction::Quit),
                KeyCode::Char('l') | KeyCode::Char('L') => return Ok(SessionAction::Logout),
                KeyCode::Up | KeyCode::Char('w') | KeyCode::Char('W') => Some((-1, 0)),
                KeyCode::Down | KeyCode::Char('s') | KeyCode::Char('S') => Some((1, 0)),
                KeyCode::Left | KeyCode::Char('a') | KeyCode::Char('A') => Some((0, -1)),
                KeyCode::Right | KeyCode::Char('d') | KeyCode::Char('D') => Some((0, 1)),
                _ => None,
            };
            let Some((dr, dc)) = movement else {
                continue;
            };
            let result = try_move(row, col, dr, dc);
            if result.moved {
                previous = Some(Position { row, col });
                row = result.row;
                col = result.col;
                move_count += 1;
            }
        }
    }
}

fn main() -> io::Result<()> {
    let app = App::new();
    let mut stdout = io::stdout();
    loop {
        let username = read_username()?;
        terminal::enable_raw_mode()?;
        execute!(stdout, Hide)?;
        let action = run_grid(&mut stdout, &app, &username)?;
        execute!(stdout, Show)?;
        terminal::disable_raw_mode()?;
        if action == SessionAction::Quit {
            break;
        }
    }
    Ok(())
}
